runtime/src/main/cpp/Memory.cpp

```diff
@@ -1747,19 +1747,13 @@ MemoryState* initMemory() {
   return memoryState;
 }
 
-void deinitMemory(MemoryState* memoryState) {
-  static int pendingDeinit = 0;
-  atomicAdd(&pendingDeinit, 1);
+bool deinitMemory(MemoryState* memoryState) {
 #if USE_GC
   bool lastMemoryState = atomicAdd(&aliveMemoryStatesCount, -1) == 0;
-  bool checkLeaks = g_checkLeaks && lastMemoryState;
+  bool checkLeaks = Kotlin_memoryLeakCheckerEnabled() && lastMemoryState;
   if (lastMemoryState) {
    garbageCollect(memoryState, true);
 #if USE_CYCLIC_GC
-   // If there are other pending deinits (rare situation) - just skip the leak checker.
```

Comment:
Can this race still happen with threads that aren't produced with `Worker.start()`?

Reply:
True, restored the check and added a comment explaining when this race may happen.